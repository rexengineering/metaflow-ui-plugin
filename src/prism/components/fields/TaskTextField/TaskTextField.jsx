import React, { useCallback } from "react";
import PropTypes from "prop-types";
import { TextField } from "@material-ui/core";
import { useField } from "formik";

function TaskTextField({ name, validateFn, ...passProps }) {
  const [
    { onChange: onChangeFormik, onBlur: onBlurFormik, ...field },
    { error, touched },
    { setError },
  ] = useField(name);

  const onChange = useCallback(
    (event) => {
      onChangeFormik(event);
      validateFn(name, event?.target?.value, setError);
    },
    [name, onChangeFormik, setError, validateFn]
  );

  const onBlur = useCallback(
    (event) => {
      onBlurFormik(event);
      validateFn(name, event?.target?.value, setError);
    },
    [name, onBlurFormik, setError, validateFn]
  );

  return (
    <TextField
      variant="filled"
      error={!!(touched && error)}
      helperText={touched && error}
      // Disabled as this is a pass-through for MUI and Formik props
      /* eslint-disable react/jsx-props-no-spreading */
      {...passProps}
      {...field}
      onChange={onChange}
      onBlur={onBlur}
      /* eslint-enable react/jsx-props-no-spreading */
    />
  );
}

TaskTextField.propTypes = {
  name: PropTypes.string.isRequired,
  validateFn: PropTypes.func,
};

TaskTextField.defaultProps = {
  validateFn: () => {},
};

export default TaskTextField;
