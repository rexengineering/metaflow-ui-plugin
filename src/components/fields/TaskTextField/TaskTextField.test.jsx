import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Formik } from "formik";
import React from "react";
import TaskTextField from ".";

describe("<TaskTextField />", () => {
  it("should render", () => {
    const { asFragment } = render(
      <Formik initialValues={{ field: "" }}>
        <TaskTextField name="field" />
      </Formik>
    );
    const initialRender = asFragment();

    expect(initialRender).not.toBeNull();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(
      <Formik initialValues={{ field: "" }}>
        <TaskTextField name="field" />
      </Formik>
    );
    const initialRender = asFragment();

    expect(initialRender).toMatchSnapshot();
  });

  it("should validate on blur", async () => {
    const validateFn = jest.fn();
    const fieldName = "field";
    render(
      <Formik initialValues={{ [fieldName]: "" }}>
        <TaskTextField
          name={fieldName}
          label="myLabel"
          validateFn={validateFn}
        />
      </Formik>
    );
    const input = screen.getByRole("textbox");

    fireEvent.blur(input);

    await waitFor(() => {
      expect(validateFn).toBeCalledTimes(1);
      expect(validateFn).toBeCalledWith(fieldName, "", expect.anything());
    });
  });

  it("should validate on change", async () => {
    const validateFn = jest.fn();
    const fieldName = "field";
    render(
      <Formik initialValues={{ [fieldName]: "" }}>
        <TaskTextField
          name={fieldName}
          label="myLabel"
          validateFn={validateFn}
        />
      </Formik>
    );
    const input = screen.getByRole("textbox");

    userEvent.type(input, "a1b2c3");

    await waitFor(() => {
      expect(validateFn).toBeCalledTimes(6);
      expect(validateFn).toHaveBeenNthCalledWith(
        6,
        fieldName,
        "a1b2c3",
        expect.anything()
      );
    });
  });
});
