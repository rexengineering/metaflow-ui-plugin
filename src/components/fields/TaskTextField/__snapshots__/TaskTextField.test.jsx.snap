// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskTextField /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <div
      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiFilledInput-input"
        name="field"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
